File: problems/1649/solution.cpp
const int MOD = 1e9 + 7;

class BIT {
private:
    vector<int> tree;
    int n;

public:
    BIT(int _n) {
        n = _n;
        tree.resize(n, 0);
    }
    
    void update(int i, int val) {
        for (; i <= n; i += i & -i) {
            tree[i] += val;
        }
    }
    
    int query(int i) {
        int sum = 0;
        
        for (; i > 0; i -= i & -i) {
            sum += tree[i];
        }
        
        return sum;
    }
};

class Solution {
public:
    int createSortedArray(vector<int>& instructions) {
        int ans = 0, n = instructions.size();
        BIT *bit = new BIT(100002);
        
        for (int i = 0; i < n; i++) {
            int lsum = bit->query(instructions[i] - 1), rsum = i - bit->query(instructions[i]);
            
            ans += min(lsum, rsum);
            ans %= MOD;
            
            bit->update(instructions[i], 1);
        }
        
        return ans % MOD;
    }
};

File: problems/1649/solution.go
package main
const MOD = 1e9 + 7

func createSortedArray(instructions []int) int {
    ans, n, bit := 0, len(instructions), NewBIT(100002)
    
    for i := 0; i < n; i++ {
        lsum, rsum := bit.Query(instructions[i] - 1), i - bit.Query(instructions[i])
        ans += min(lsum, rsum)
        ans %= MOD
        bit.Update(instructions[i], 1)
    }
    
    return ans % MOD
}

func min(a, b int) int {
    if a < b {
        return a
    }
    
    return b
}

type BIT struct {
	tree []int
	size int
}

func NewBIT(size int) *BIT {
	return &BIT{
		make([]int, size),
		size,
	}
}

func (this *BIT) Update(index, val int) {
	for ; index <= this.size; index += index & -index {
		this.tree[index] += val
	}
}

func (this *BIT) Query(index int) int {
	sum := 0

	for ; index > 0; index -= index & -index {
		sum += this.tree[index]
	}

	return sum
}